pennylane/devices/qubit/preprocess.py

```diff
@@ -87,13 +96,30 @@ def expand_fn(circuit, max_expansion=10):
     if any(isinstance(o, MidMeasureMP) for o in circuit.operations):
         circuit = qml.defer_measurements(circuit)
 
-    if not all(_stopping_condition(op) for op in circuit.operations):
-        circuit = circuit.expand(depth=max_expansion, stop_at=_stopping_condition)
```

Comment:
i feel a little iffy on just.. re-implementing expansion because it's not quite compatible with the new device. the change looks good, but I wanna be a bit more confident in the need for this.

Reply:
Some time in the future, I want to add a class into a `devices.preprocessing` submodule that looks like:
```
class DecomposeTillSupportedOperation:

    def __init__(self, supported_operations: set[str]):

     def __call__(self, circuit: Union[qml.tape.QuantumScript, Iterable[qml.tape.QuantumScript]]):
```

but it would have a look more external-developer facing design considerations than what we are doing here.

Reply:
imo this PR is re-developing a big aspect of pennylane and probably requires some more discussion

Reply:
The entire point is to avoid having to modify or redesign `tape.expand`, because that would touch too many other things. This is simply an internal implementation detail of an experimental feature.  We are separating out a lot of this code so we won't be limited by decisions made in completely independent parts of the code base.

Just because we implement this here, does not mean we have to use this implementation any where else.  We can *choose* to, but that decision can be made on a case-by-case basis.

We're not telling any other piece of code *solve this problem this way*. If anything, we are reducing coupling between implementation decisions.

Reply:
I see your point. I guess what I'm saying is that I am enthusiastic about this new implementation, and I anticipate that it will be re-used in the future, so I'd rather have exist as stand-alone utility code, rather than have to come back and factor it into its own file in the future when another device wants it. For now I'll just assume/pretend that won't happen 😄

Reply:
> I see your point. I guess what I'm saying is that I am enthusiastic about this new implementation, and I anticipate that it will be re-used in the future, so I'd rather have exist as stand-alone utility code, rather than have to come back and factor it into its own file in the future when another device wants it. For now I'll just assume/pretend that won't happen 

@timmysilv are you just suggesting we factor this section out into a utility function in this PR? Isn't this something fairly reasonable to do.

Reply:
This is an option for the future, but I would argue it's not something we need to deal with right now.  We can start using this pattern in other places if it makes sense.  Once we start using the pattern a lot, we can re-examine whether or not we want to extract any code, and what that extracted utility function could look like.

Reply:
@trbromley yes, buuuut my generalized opinion on DRY code is as soon as you copy-paste code, extract it (in other words, if you do the same thing even twice). So I'd actually like to rescind my former comment in the context of this PR! If we need it again, we'll talk. lgtm 👌